### PR TITLE
Use RazorViewAttribute \ RazorPageAttribute for view discovery

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/CompiledViewManifest.cs [hunk 1]
```diff
@@ -13,13 +11,7 @@ public static class CompiledViewManfiest
     {
         public static readonly string PrecompiledViewsAssemblySuffix = ".PrecompiledViews";
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/CompiledViewManifest.cs]: There's a couple of types referenced by precompilation But we need to switch to attributes first to get precompilation tests to work once we delete codes from it. This is first pass. Once this is in, we can delete a bunch of code from precompilation and then follow up with removing code from this repo.

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/CompiledViewManifest.cs [hunk 2]
```diff
@@ -12,13 +11,7 @@ public static class CompiledViewManfiest
     {
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/CompiledViewManifest.cs]: We should totally rename this to its correct spelling for 2.0

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationParts/CompiledPageFeatureProvider.cs [hunk 3]
```diff
@@ -45,30 +46,39 @@ public void PopulateFeature(IEnumerable<ApplicationPart> parts, ViewsFeature fea
...
+            if (assemblyPart == null)
+            {
+                throw new ArgumentNullException(nameof(assemblyPart));
+            }
 
-                yield return viewDescriptor;
+            var featureAssembly = CompiledViewManfiest.GetFeatureAssembly(assemblyPart);
+            if (featureAssembly != null)
+            {
+                return featureAssembly.GetCustomAttributes<RazorPageAttribute>();;
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationParts/CompiledPageFeatureProvider.cs]: ;; => ;

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/CompiledViewManifest.cs [hunk 4]
```diff
@@ -12,13 +11,7 @@ public static class CompiledViewManfiest
...
         public static readonly string PrecompiledViewsAssemblySuffix = ".PrecompiledViews";
 
-        public static Type GetManifestType(AssemblyPart assemblyPart, string typeName)
-        {
-            var assembly = GetFeatureAssembly(assemblyPart);
-            return assembly?.GetType(typeName);
-        }
-
-        private static Assembly GetFeatureAssembly(AssemblyPart assemblyPart)
+        public static Assembly GetFeatureAssembly(AssemblyPart assemblyPart)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/CompiledViewManifest.cs]: Unit test?

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ViewsFeatureProvider.cs [hunk 5]
```diff
@@ -57,19 +53,25 @@ public void PopulateFeature(IEnumerable<ApplicationPart> parts, ViewsFeature fea
...
+            {
+                throw new ArgumentNullException(nameof(assemblyPart));
+            }
+
+            var featureAssembly = CompiledViewManfiest.GetFeatureAssembly(assemblyPart);
+            if (featureAssembly != null)
             {
-                return (ViewInfoContainer)Activator.CreateInstance(type);
+                return featureAssembly.GetCustomAttributes<RazorViewAttribute>()
+                    .Where(attribute => attribute.GetType() == typeof(RazorViewAttribute));
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ViewsFeatureProvider.cs]: Any harm in letting sub-types fit in here?

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/CompiledViewManifest.cs (hunk 4 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/CompiledViewManifest.cs]: We have tests via `ViewsFeatureProvider`: https://github.com/aspnet/Mvc/blob/dev/test/Microsoft.AspNetCore.Mvc.Razor.Test/Compilation/ViewsFeatureProviderTest.cs#L94

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ViewsFeatureProvider.cs (hunk 5 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ViewsFeatureProvider.cs]: Maybe not. We can remove the other page feature provider and let this type do the discovery for pages and views alike.

**Comment by pranavkm** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationParts/CompiledPageFeatureProvider.cs [hunk 6]
```diff
@@ -29,56 +22,5 @@ public class CompiledPageFeatureProvider : IApplicationFeatureProvider<ViewsFeat
 
         private static readonly string FullyQualifiedManifestTypeName =
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationParts/CompiledPageFeatureProvider.cs]: This doesn't seem to be used.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageApplicationModelProvider.cs [hunk 7]
```diff
@@ -78,8 +80,33 @@ private void EnsureCache()
...
+                .OfType<AssemblyPart>()
+                .SelectMany(GetAttributes);
+        }
+
+        private static IEnumerable<RazorPageAttribute> GetAttributes(AssemblyPart assemblyPart)
+        {
+            var featureAssembly = CompiledViewManfiest.GetFeatureAssembly(assemblyPart);
+            if (featureAssembly != null)
+            {
+                return featureAssembly.GetCustomAttributes<RazorPageAttribute>(); ;
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageApplicationModelProvider.cs]: lol `; ;`

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/CompiledViewManifest.cs (hunk 2 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/CompiledViewManifest.cs]: Should have something to track this so it doesn't fall off the radar

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ViewsFeatureProvider.cs (hunk 5 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ViewsFeatureProvider.cs]: Yeah this should probably be `OfType`

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/CompiledViewManifest.cs (hunk 2 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/CompiledViewManifest.cs]: I'm going to get this in and clean up precompilation and follow up on this today. 😄

**Pull request closed without merging**